configserver/src/main/java/com/yahoo/vespa/config/server/application/DefaultClusterReindexingStatusClient.java

```diff
@@ -0,0 +1,150 @@
+            @Override public void completed(SimpleHttpResponse result) { responsePromise.complete(result); }
+            @Override public void failed(Exception ex) { responsePromise.completeExceptionally(ex); }
+            @Override public void cancelled() { responsePromise.cancel(false); }
+        });
+        return responsePromise.handleAsync((response, error) ->  {
+            if (response != null) {
+                return uncheck(() -> toClusterReindexing(response));
+            } else {
+                throw throwUnchecked(new IOException(String.format("For '%s': %s", uri, error.getMessage()), error));
+            }
+        }, executor);
+    }
+
+    private static ClusterReindexing toClusterReindexing(SimpleHttpResponse response) throws IOException {
+        if (response.getCode() != HttpStatus.SC_OK) throw new IOException("Expected status code 200, got " + response.getCode());
+        if (response.getBody() == null) throw new IOException("Response has no content");
+        return toClusterReindexing(response.getBodyBytes());
+    }
+
+    private static ClusterReindexing toClusterReindexing(byte[] requestBody) throws IOException {
+        JsonNode jsonNode = mapper.readTree(requestBody);
+        Map<String, ClusterReindexing.Status> documentStatuses = new HashMap<>();
+        for (JsonNode statusJson : jsonNode.get("status")) {
+            String type = statusJson.get("type").textValue();
+            Instant startedMillis = Instant.ofEpochMilli(statusJson.get("startedMillis").longValue());
+            Instant endedMillis = Instant.ofEpochMilli(statusJson.get("endedMillis").longValue());
+            String progressToken = statusJson.get("progress").textValue();
+            ClusterReindexing.State state = ClusterReindexing.State.fromString(statusJson.get("state").textValue());
+            String message = statusJson.get("message").textValue();
+            documentStatuses.put(type, new ClusterReindexing.Status(startedMillis, endedMillis, state, message, progressToken));
+        }
+        return new ClusterReindexing(documentStatuses);
+    }
+
+    private static int getStatePort(ServiceInfo service) {
+        return service.getPorts().stream()
+                .filter(port -> port.getTags().contains("state"))
+                .map(PortInfo::getPort)
+                .findAny()
+                .orElseThrow(() -> new IllegalStateException("Cluster controller container has no container port"));
+    }
+
+    private static Map<ClusterId, List<ServiceInfo>> clusterControllerClusters(ModelResult application) {
+        return application.getModel().getHosts().stream()
+                .flatMap(host -> host.getServices().stream())
+                .filter(service -> service.getServiceType().equals(CLUSTERCONTROLLER_CONTAINER.serviceName))
+                .collect(Collectors.groupingBy(service -> new ClusterId(service.getProperty("clustername").get())));
```

Comment:
Whenever I see `property` and `name` in the same sentence I get really excited :)))